File: frontend/src/api/stageData.ts
import { apiRequest, downloadAuthenticatedFile } from "./auth";

export type BusinessDomain = "ENGINEERING" | "PRODUCTION";
export type TestStage = "CP" | "FT";

export interface PageResult<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

export interface StagePageRequest {
  page: number;
  page_size: number;
  factory_code?: string;
  status?: string;
  product_name?: string;
  lot_id?: string;
  from_utc?: string;
  to_utc?: string;
}

export interface StageUploadRow {
  import_batch_id: number;
  sequence_no: number;
  receipt_id: number;
  source_file_id: number;
  original_file_name: string;
  extension: string;
  size_bytes: number;
  factory_code: string;
  upload_time_utc: string;
  completion_time_utc: string | null;
  uploader_login: string;
  uploader_name: string;
  status: string;
  latest_job_id: number | null;
  error_code: string | null;
  error_message: string | null;
  action_required: "LOT_ID" | null;
  queue_age_seconds?: number | null;
}

export interface StageInputRequest {
  input_request_id: number;
  source_file_id: number;
  original_file_name: string;
  current_value: null;
}

export interface StageInputRequests {
  import_batch_id: number;
  status: string;
  field_code: "LOT_ID";
  prompt: string;
  latest_job_id: number | null;
  requests: StageInputRequest[];
}

export interface ResolveStageInputRequestsPayload {
  resolutions: Array<{ input_request_id: number; lot_id: string }>;
  reason: string;
}

export interface ResolveStageInputRequestsResult {
  import_batch_id: number;
  job_id: number;
  status: "QUEUED";
}

export interface StageResultRow {
  result_summary_id: number;
  import_batch_id: number;
  data_name: string;
  product_name: string | null;
  lot_id: string | null;
  wafer_count: number | null;
  factory_code: string;
  test_item_count: number | null;
  unit_count: number | null;
  pass_count: number | null;
  yield_rate: number | null;
  status: string;
  data_type: string;
  dataset_id: number | null;
  dataset_version_no: number | null;
  job_id?: number | null;
  created_at_utc: string;
}

export interface FormalSourceRoot {
  code: string;
  name: string;
  test_stage: TestStage;
  factory_code: string;
  allowed_suffixes: string[];
  purpose: "FORMAL_IMPORT";
  business_domains: BusinessDomain[];
  available: boolean;
}

export interface FormalSourceDirectory {
  name: string;
  relative_path: string;
  direct_file_count: number;
  direct_total_bytes: number;
}

export interface FormalDirectoryListing {
  root_code: string;
  current_relative_path: string;
  parent_relative_path: string | null;
  directories: FormalSourceDirectory[];
}

export interface FormalSourceManifestPreview {
  root_code: string;
  relative_path: string;
  mode: string;
  recursive: boolean;
  file_count: number;
  total_bytes: number;
  sha: string;
  allowed_suffixes: string[];
}

const stageBase = (businessDomain: BusinessDomain, testStage: TestStage) =>
  `/api/v1/${businessDomain.toLowerCase()}/${testStage.toLowerCase()}`;

export const listStageUploads = (businessDomain: BusinessDomain, testStage: TestStage) =>
  apiRequest<StageUploadRow[]>(`${stageBase(businessDomain, testStage)}/uploads`);

export const listStageResults = (businessDomain: BusinessDomain, testStage: TestStage) =>
  apiRequest<StageResultRow[]>(`${stageBase(businessDomain, testStage)}/results`);

const stagePageQuery = (request: StagePageRequest) => {
  const query = new URLSearchParams({
    page: String(request.page),
    page_size: String(request.page_size),
  });
  for (const key of ["factory_code", "status", "product_name", "lot_id", "from_utc", "to_utc"] as const) {
    const value = request[key]?.trim();
    if (value) query.set(key, value);
  }
  return query;
};

export const listStageUploadsPage = (
  businessDomain: BusinessDomain,
  testStage: TestStage,
  request: StagePageRequest,
) => apiRequest<PageResult<StageUploadRow>>(
  `${stageBase(businessDomain, testStage)}/uploads/page?${stagePageQuery(request)}`,
);

export const listStageResultsPage = (
  businessDomain: BusinessDomain,
  testStage: TestStage,
  request: StagePageRequest,
) => apiRequest<PageResult<StageResultRow>>(
  `${stageBase(businessDomain, testStage)}/results/page?${stagePageQuery(request)}`,
);

export const listFormalSourceRoots = (
  businessDomain: BusinessDomain,
  testStage: TestStage,
  factoryCode: string,
) => {
  const query = new URLSearchParams({ factory_code: factoryCode });
  return apiRequest<FormalSourceRoot[]>(
    `${stageBase(businessDomain, testStage)}/source-roots?${query}`,
  );
};

export const listFormalSourceDirectories = (
  businessDomain: BusinessDomain,
  testStage: TestStage,
  factoryCode: string,
  rootCode: string,
  relativePath = ".",
) => {
  const query = new URLSearchParams({
    factory_code: factoryCode,
    relative_path: relativePath,
  });
  return apiRequest<FormalDirectoryListing>(
    `${stageBase(businessDomain, testStage)}/source-roots/${encodeURIComponent(rootCode)}/directories?${query}`,
  );
};

export const previewFormalSourceManifest = (
  businessDomain: BusinessDomain,
  testStage: TestStage,
  factoryCode: string,
  rootCode: string,
  relativePath = ".",
) => {
  const query = new URLSearchParams({
    factory_code: factoryCode,
    relative_path: relativePath,
  });
  return apiRequest<FormalSourceManifestPreview>(
    `${stageBase(businessDomain, testStage)}/source-roots/${encodeURIComponent(rootCode)}/manifest-preview?${query}`,
  );
};

export function uploadStageData(
  businessDomain: BusinessDomain,
  testStage: TestStage,
  files: File[],
  factoryCode: string,
  remark?: string,
  sourceRootCode?: string,
  sourceRelativePath?: string,
  sourceManifestMode?: string,
  sourceManifestSha256?: string,
) {
  const body = new FormData();
  files.forEach((file) => body.append("files", file));
  body.append("factory_code", factoryCode);
  if (sourceRootCode?.trim()) {
    body.append("source_root_code", sourceRootCode.trim());
    body.append("source_relative_path", sourceRelativePath?.trim() || ".");
    if (sourceManifestMode?.trim()) {
      body.append("source_manifest_mode", sourceManifestMode.trim());
    }
    if (sourceManifestSha256?.trim()) {
      body.append("source_manifest_sha256", sourceManifestSha256.trim());
    }
  }
  if (remark) body.append("remark", remark);
  return apiRequest<{
    import_batch_id: number;
    job_id: number;
    status: "QUEUED";
    input_mode: "WEB_UPLOAD" | "SOURCE_CATALOG";
    business_domain: BusinessDomain;
    test_stage: TestStage;
    cleaner_release: {
      cleaner_release_id: number;
      cleaner_code: string;
      cleaner_version: string;
    };
  }>(`${stageBase(businessDomain, testStage)}/uploads`, { method: "POST", body });
}

export const reprocessStageBatch = (businessDomain: BusinessDomain, testStage: TestStage, importBatchId: number) =>
  apiRequest<{ import_batch_id: number; status: string }>(`${stageBase(businessDomain, testStage)}/uploads/${importBatchId}/reprocess`, { method: "POST" });

export const getStageInputRequests = (businessDomain: BusinessDomain, testStage: TestStage, importBatchId: number) =>
  apiRequest<StageInputRequests>(`${stageBase(businessDomain, testStage)}/uploads/${importBatchId}/input-requests`);

export const resolveStageInputRequests = (
  businessDomain: BusinessDomain,
  testStage: TestStage,
  importBatchId: number,
  payload: ResolveStageInputRequestsPayload,
) => apiRequest<ResolveStageInputRequestsResult>(
  `${stageBase(businessDomain, testStage)}/uploads/${importBatchId}/input-requests/resolve`,
  { method: "POST", body: JSON.stringify(payload) },
);

export async function downloadStageUploadFile(businessDomain: BusinessDomain, testStage: TestStage, importBatchId: number, receiptId: number, fileName: string) {
  return downloadAuthenticatedFile(
    `${stageBase(businessDomain, testStage)}/uploads/${importBatchId}/files/${receiptId}/download`,
    fileName,
  );
}

File: backend/app/domain/stage_data.py
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Literal, Protocol

from app.domain.auth import Principal


@dataclass(frozen=True, slots=True)
class StoredUpload:
    original_name: str
    path: Path
    size_bytes: int
    sha256: str
    source_metadata: dict[str, object] | None = None


@dataclass(frozen=True, slots=True)
class FormalSourceManifestPreview:
    root_code: str
    relative_path: str
    mode: str
    recursive: bool
    file_count: int
    total_bytes: int
    sha: str
    allowed_suffixes: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class StageUploadRow:
    import_batch_id: int
    sequence_no: int
    receipt_id: int
    original_file_name: str
    extension: str
    size_bytes: int
    factory_code: str
    upload_time_utc: str
    completion_time_utc: str | None
    uploader_login: str
    uploader_name: str
    status: str
    source_file_id: int
    latest_job_id: int | None = None
    error_code: str | None = None
    error_message: str | None = None
    action_required: Literal["LOT_ID"] | None = None
    is_duplicate_receipt: bool = False
    can_manage: bool = False
    can_download_source: bool = False


@dataclass(frozen=True, slots=True)
class BatchFileInfo:
    receipt_id: int
    original_file_name: str
    storage_uri: str
    source_file_id: int
    expected_sha256: str | None
    lot_id_override: str | None = None
    is_duplicate_receipt: bool = False


@dataclass(frozen=True, slots=True)
class BatchInfo:
    import_batch_id: int
    factory_code: str
    status: str
    files: tuple[BatchFileInfo, ...]
    can_manage: bool = False


@dataclass(frozen=True, slots=True)
class WorkerBatchInfo:
    import_batch_id: int
    business_domain: str
    test_stage: str
    factory_code: str
    files: tuple[BatchFileInfo, ...]


@dataclass(frozen=True, slots=True)
class StageResultRow:
    result_summary_id: int
    import_batch_id: int
    data_name: str
    product_name: str | None
    lot_id: str | None
    wafer_count: int | None
    factory_code: str
    test_item_count: int | None
    unit_count: int | None
    pass_count: int | None
    yield_rate: float | None
    status: str
    data_type: str
    dataset_id: int | None
    dataset_version_no: int | None
    created_at_utc: str
    can_manage: bool = False
    uploader_login: str = ""
    uploader_name: str = ""


class StageDataService(Protocol):
    def register_upload(
        self,
        principal: Principal,
        business_domain: str,
        test_stage: str,
        factory_code: str,
        files: Sequence[StoredUpload],
        remark: str | None,
    ) -> int: ...

    def mark_processing(self, batch_id: int, principal: Principal) -> int: ...

    def mark_queued(self, batch_id: int) -> None: ...

    def mark_failed(self, batch_id: int, job_id: int, message: str) -> None: ...

    def record_result(self, batch_id: int, job_id: int, result: dict) -> None: ...

    def list_uploads(
        self, principal: Principal, business_domain: str, test_stage: str
    ) -> tuple[StageUploadRow, ...]: ...

    def list_results(
        self, principal: Principal, business_domain: str, test_stage: str
    ) -> tuple[StageResultRow, ...]: ...

    def get_batch_info(
        self, principal: Principal, business_domain: str, test_stage: str, batch_id: int
    ) -> BatchInfo | None: ...

    def archive_previous_results(self, batch_id: int) -> None: ...
